File: unmastered/ctci_problems/linked_lists/2.5.js
var LinkedList = require('../../../mastered/data_structures/linked_list').LinkedList;
var Node = require('../../../mastered/data_structures/linked_list').Node;

/*  greater of m + n  + 1 space.  linear time too
set list1 and list2 to their respective nexts
set carriedOver to 0
create result list, in which we always have a pointer to the node before tail, so we can create new nodes when necessary
while at least one pointer is not at its tail or carriedOver is not zero, do the following:
	set currPlaceValue to be list1s currNode's value + list2 currNodes's value + carriedOver
	set carriedOver to currPlaceValue / 10 rounded down
	create a new node with value currPlaceValue % 10
	Increment those pointers if applicable
return the new list
*/

/* FORWARD ORDER

*/

function sumListsForwardOrder(list1, list2){ //FORWARD ORDER IMPLEMENTATION

}



// let addend1 = new LinkedList();
// let addend2 = new LinkedList();

// addend1.addNode(7);
// addend1.addNode(7);
// addend1.addNode(9);

// addend2.addNode(2);
// addend2.addNode(2);
// addend2.addNode(3)

// let res = sumListsForwardOrder(addend1, addend2);
// res.print();



// function sumListsForwardOrder(list1, list2){ //FORWARD ORDER IMPLEMENTATION
// 	let res = new LinkedList();
// 	let additionalPlace = sumListsRecursHelper(list1.head.next, list2.head.next, res);
// 	if (additionalPlace > 0){
// 		addNode(res, additionalPlace);
// 	}
// 	return res;
// }

// function sumListsRecursHelper(list1Ptr, list2Ptr, resList){
// 	if (list1Ptr.val === undefined && list2Ptr.val === undefined){
// 		return 0;
// 	}
// 	let thisPlaceValue = list1Ptr.val || 0;
// 	thisPlaceValue += list2Ptr.val || 0;
// 	let next1 = list1Ptr.val === undefined ? list1Ptr : list1Ptr.next;
// 	let next2 = list2Ptr.val === undefined ? list2Ptr : list2Ptr.next;
// 	thisPlaceValue += sumListsRecursHelper(next1, next2, resList);
// 	let nextPlaceGreater = Math.floor(thisPlaceValue / 10);
// 	let thisPlace = thisPlaceValue % 10;
// 	addNode(resList, thisPlace);
// 	return nextPlaceGreater;
// }

// function addNode(list, nodeValue){
// 	let newNode = new Node(nodeValue);
// 	newNode.next = list.head.next;
// 	list.head.next = newNode;
// }

File: unmastered/ctci_problems/linked_lists/2.5.py
"""
	Only do forward order list

"""














































def addForwardLists(l1, l2):
	res = LinkedList();
	l1PlaceOut = l1.countPlaces();
	l2PlaceOut = l2.countPlaces();
	l1Ptr = l1.head.next
	l2Ptr = l2.head.next

	while (l1PlaceOut > 0 or l2PlaceOut > 0):
		if l1PlaceOut > l2PlaceOut:

		elif l2PlaceOut > l1PlaceOut:

		else:
			thisSum = l1Ptr.val + l2Ptr.val
			thisPlaceNode = Node(thisSum % 10);

class Node(object):
	def __init__(self, val):
		self.val = val;
		self.next = None;
		self.prev = None;

class LinkedList(object):
	def __init__(self):
		self.head = Node(None);
		self.tail = Node(None);
		self.head.next = self.tail;
		self.tail.prev = self.head;
		self.head.prev = None;
		self.tail.next = None;

	def addNode(self, node):
		self.tail.prev.next = node;
		node.prev = self.tail.prev;
		self.tail.prev = node;
		node.next = self.tail;

	def prnt(self):
		currPtr = self.head.next
		while (currPtr != self.tail):
			print currPtr.val;
			currPtr = currPtr.next;

	def countPlaces(self):
		currPtr = self.head.next;
		count = 0;
		while (currPtr != self.tail):
			count++;
			currPtr = currPtr.next;
		return count;

def addLists(l1, l2):
	sum = LinkedList();
	l1Ptr = l1.head.next;
	l2Ptr = l2.head.next;
	remainder = 0;
	while (l1Ptr != l1.tail or l2Ptr != l2.tail):
		thisSum = remainder
		if (l1Ptr != l1.tail):
			thisSum += l1Ptr.val
			l1Ptr = l1Ptr.next;
		if (l2Ptr != l2.tail):
			thisSum += l2Ptr.val
			l2Ptr = l2Ptr.next
		thisDigitNode = Node(thisSum % 10)
		remainder = thisSum / 10
		sum.addNode(thisDigitNode);
	if remainder > 0:
		sum.addNode(remainder);
	return sum;

def test():
	l1 = LinkedList();
	l2 = LinkedList();

	l1.addNode(Node(7))
	l1.addNode(Node(1))
	l1.addNode(Node(6))
	l1.addNode(Node(1))

	l2.addNode(Node(5))
	l2.addNode(Node(9))
	l2.addNode(Node(2))

	res = addLists(l1, l2);
	res.prnt()

test();